src/provision/freedom-module.js

```diff
@@ -0,0 +1,351 @@
+  this.dispatch("status", {
+    "code": code,
+    "message": STATUS_CODES[code]
+  });
+};
+
+/*
+ * Generates an RSA keypair using forge
+ * @return {Object.<String,String>} public and private SSH keys
+ */
+Provisioner.prototype._generateKeyPair = function() {
+  "use strict";
+  var pair = forge.pki.rsa.generateKeyPair({bits: 1024, e: 0x10001});  // TODO: use async
+  var publicKey = forge.ssh.publicKeyToOpenSSH(pair.publicKey, 'info@uproxy.org');
+  var privateKey = forge.ssh.privateKeyToOpenSSH(pair.privateKey, '');
+  return { public: publicKey, private: privateKey };
+}
+
+/**
+ * Initiates a Digital Ocean oAuth flow
+ * @return {Promise.<Object>} oAuth response from Digital Ocean
+ *  {
+ *    access_token: "..",
+ *    expires_in: "..",
+ *    state: "..",
+ *    token_type: ".."
+ *  }
+ **/
+Provisioner.prototype._doOAuth = function() {
+  return new Promise(function(resolve, reject) {
+    var oauth = freedom["core.oauth"]();
+
+    this._sendStatus("OAUTH_INIT");
+    oauth.initiateOAuth(REDIRECT_URIS).then(function(obj) {
+      var url = "https://cloud.digitalocean.com/v1/oauth/authorize?" +
+                "client_id=c16837b5448cd6cf2582d2c2f767cfb7d11844ec395a91b43f26ca72513416c8&" +
+                "response_type=token&" +
+                "redirect_uri=" + encodeURIComponent(obj.redirect) + "&" +
+                "state=" + encodeURIComponent(obj.state) + "&" +
+                "scope=read%20write";
+      return oauth.launchAuthFlow(url, obj);
+    }).then(function(responseUrl) {
+      var query = responseUrl.substr(responseUrl.indexOf('#') + 1),
+        param,
+        params = {},
+        keys = query.split('&'),
+        i = 0;
+
+      for (i = 0; i < keys.length; i += 1) {
+        param = keys[i].substr(0, keys[i].indexOf('='));
+        params[param] = keys[i].substr(keys[i].indexOf('=') + 1);
+      }
+
+      this._sendStatus("OAUTH_COMPLETE");
+      resolve(params);
+    }.bind(this)).catch(function(err) {
+      console.error("oauth error: " + JSON.stringify(err));
+      this._sendStatus("OAUTH_ERROR");
+      reject(err)
+    }.bind(this));
+  }.bind(this));
+};
+
+/**
+ * Try to retrieve SSH keys from storage.
+ * If not found, generate new ones and store
+ * @param {String} name of the key (usually same as name of VM later)
+ * @return {Object}
+ * {
+ *    public: "...",
+ *    private: "..."
+ * }
+ **/
+Provisioner.prototype._getSshKey = function(name) {
+  var storage = freedom["core.storage"]();
+  return new Promise(function(resolve, reject) {
+    var result = {};
+
+    Promise.all([
+      storage.get("DigitalOcean-" + name + "-PublicKey"),
+      storage.get("DigitalOcean-" + name + "-PrivateKey")
+    ]).then(function(val) {
+      if (val[0] === null ||
+         val[1] === null) {
+        result = this._generateKeyPair();
+        storage.set("DigitalOcean-" + name + "-PublicKey", result.public);
+        storage.set("DigitalOcean-" + name + "-PrivateKey", result.private);
```

Comment:
do we have a bug to be able to store at least the private key encrypted with a passphrase, so that the user needs to decrypt it at time of use?